acceptance/ui/features/api/ui.rb

```diff
@@ -0,0 +1,231 @@
+require "capybara"
+require "capybara/rspec"
+require_relative "../pages/applications"
+require_relative "../pages/hosts"
+require_relative "../pages/pools"
+require_relative "../pages/service"
+require_relative "../pages/servicesmap"
+require_relative "../pages/user"
+
+#
+# Contains the list of page objects to be returned.  Login is only
+# done if the session is invalid.
+#
+class UI
+    include ::RSpec::Matchers
+    include ::Capybara::DSL
+
+    def initialize()
+        @pages = {
+            applications: Applications.new,
+            hosts: Hosts.new,
+            login: nil,
+            pools: Pools.new,
+            services: Service.new,
+            servicesMap: ServicesMap.new,
+            user: User.new,
+        }
+    end
+
+    # Ensures that we've logged in.  Subsequent calls
+    # will use the same page and not login again.
+    def login()
+        if self.LoginPage == nil || !verify_login?()
+            puts "  \e[33m** Logging in to the UI **\e[0m"
+            login_as(applicationUserID(), applicationPassword())
+        end
+    end
+
+    def login_as(username, password)
+        @pages[:login] = nil # Reset the login data.
+        do_login(username, password)
+    end
+
+    def visit_login_page()
+        @pages[:login] = Login.new if self.LoginPage == nil
+        visit_login_page_impl(self.LoginPage)
+    end
+
+    def LoginPage
+        return @pages[:login]
+    end
+
+    def HostsPage
+        suppress_deploy_wizard()
+        return @pages[:hosts]
+    end
+
+    def ApplicationsPage
+        suppress_deploy_wizard()
+        return @pages[:applications]
+    end
+
+    def ServicesMapPage
+        suppress_deploy_wizard()
+        return @pages[:servicesMap]
+    end
+
+    def PoolsPage
+        suppress_deploy_wizard()
+        return @pages[:pools]
+    end
+
+    def ServicesPage
+        suppress_deploy_wizard()
+        return @pages[:services]
+    end
+
+    def UserPage
+        suppress_deploy_wizard()
+        return @pages[:user]
+    end
+
+    # Look up the table data for the given port and remove it using
+    # the UI.
+    def remove_publicendpoint_port_json(name)
```

Comment:
These methods related to publicendpoints seem out of place in this class since it is mostly a 'container' class for the different pages, as well has "global" methods like the login stuff. 

How about creating a Page object for the Add-public-endpoints dialog and adding these methods there? Alternatively,  you could move these endpoint-related methods to the Service page object.